src/basic/terminal-util.c

```diff
@@ -1206,30 +1208,52 @@ bool terminal_is_dumb(void) {
         return getenv_terminal_is_dumb();
 }
 
-bool colors_enabled(void) {
-
-        /* Returns true if colors are considered supported on our stdout. For that we check $SYSTEMD_COLORS first
-         * (which is the explicit way to turn colors on/off). If that didn't work we turn colors off unless we are on a
-         * TTY. And if we are on a TTY we turn it off if $TERM is set to "dumb". There's one special tweak though: if
-         * we are PID 1 then we do not check whether we are connected to a TTY, because we don't keep /dev/console open
-         * continuously due to fear of SAK, and hence things are a bit weird. */
-
-        if (cached_colors_enabled < 0) {
-                int val;
-
-                val = getenv_bool("SYSTEMD_COLORS");
-                if (val >= 0)
-                        cached_colors_enabled = val;
-
+color_mode get_color_mode(void) {
+
+        /* Returns the mode used to choose output colors. The possible modes are COLOR_OFF for no colors,
+         * COLOR_16 for only the base 16 ANSI colors, COLOR_256 for more colors and COLOR_ON for unrestricted
+         * color output. For that we check $SYSTEMD_COLORS first (which is the explicit way to
+         * change the mode). If that didn't work we turn colors off unless we are on a TTY. And if we are on a TTY
+         * we turn it off if $TERM is set to "dumb". There's one special tweak though: if we are PID 1 then we do not
+         * check whether we are connected to a TTY, because we don't keep /dev/console open continuously due to fear
+         * of SAK, and hence things are a bit weird. */
+
+        if (cached_color_mode < 0) {
+                char *e;
+
+                e = getenv("SYSTEMD_COLORS");
+                if (e) {
+                        if (streq(e, "16"))
+                                cached_color_mode = COLOR_16;
+                        else if (streq(e, "256"))
+                                cached_color_mode = COLOR_256;
+                        else if (parse_boolean(e))
+                                cached_color_mode = COLOR_ON;
+                        else if (!parse_boolean(e))
+                                cached_color_mode = COLOR_OFF;
+                        else
+                                cached_color_mode = _COLOR_INVALID;
```

Comment:
It is not necessary to parse the string twice. Moreover, please also fallback when an invalid string is set to `SYSTEMD_COLORS` environment variable.

How about to introduce, say,  `parse_systemd_colors()` or something:
```
static ColorMode parse_systemd_colors(void) {
        const char *e;
        int r;

        e = getenv("SYSTEMD_COLORS");
        if (!e)
                return _COLOR_INVALID;
        r = parse_boolean(e);
        if (r >= 0)
                return r > 0 ? COLOR_ON : COLOR_OFF;
        if (streq(e, "16"))
                return COLOR_16;
        if (streq(e, "256"))
                return COLOR_256;
        return _COLOR_INVALID;
}

COLOR_MODE get_color_mode(void) {
        if (cached_color_mode >= 0)
                return cached_color_mode;

        cached_color_mode = parse_systemd_colors();
        if (cached_color_mode >= 0)
                return cached_color_mode;
        
        if (getenv("NO_COLOR"))
                return cached_color_mode = COLOR_OFF;
        ...
}
```

Reply:
Implemented something similar: I think the boolean case must be handled later, because it will parse 16 and 256 as true.